Txn: move some commands functionality around.

nrc opened the pull request:
###  What have you changed?

Following up from earlier refactoring, this PR moves some command functionality around to make things a bit tidier. This is mostly just to move things closer to where there are used/belong. The most interesting change is moving the decoding of protobufs into `Command` objects from server/service/kv to storage/txn/commands, the motivation is that the kv service should be fairly minimal and not have knowledge of storage internals.

In addition I remove one unused command.

###  What is the type of the changes?

Engineering (engineering change which doesn't change any feature or fix any issue)


PTAL @MyonKeminta @youjiali1995

zhangjinpeng1987 commented on src/storage/txn/commands.rs:
```diff
@@ -1,60 +1,212 @@
+
+impl From<BatchRollbackRequest> for Command {
+    fn from(mut req: BatchRollbackRequest) -> Command {
+        let keys = req.get_keys().iter().map(|x| Key::from_raw(x)).collect();
```
Please add a blank line behind.

nrc replied:
Sorry, I'm not clear where you want the blank line. Could you explain why?

zhangjinpeng1987 commented on src/storage/txn/commands.rs:
```diff
@@ -1,60 +1,212 @@
+    fn from(mut req: ScanLockRequest) -> Command {
+        ScanLock::new(
+            req.get_max_version().into(),
+            &req.take_start_key(),
```
I prefer to use `Key::from_raw(req.get_start_key())` here to keep consistency with other Command. Maybe
```
let start_key = if !req.get_start_key().is_empty() {
  Some(Key::from_raw(req.get_start_key()))
} else {
  None
};
```

zhangjinpeng1987 commented on src/storage/txn/commands.rs:
```diff
@@ -99,6 +251,61 @@ impl Prewrite {
+        )
+    }
+
+    pub fn with_context(
```
Seems this function is only used in test, can we add `[cfg(test)]`?

nrc replied:
No, it is used by the test crates and `cfg(test)` is not propagated upstream.

zhangjinpeng1987 commented on src/storage/mod.rs:
```diff
@@ -511,34 +503,15 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
-            callback
-        );
+    pub fn prewrite(&self, cmd: Command, callback: Callback<Vec<Result<()>>>) -> Result<()> {
+        if let CommandKind::Prewrite(commands::Prewrite { mutations, .. }) = &cmd.kind {
```
The cmd is required as Prewrite type, can we use a strong constraint to guarantee it here?

nrc replied:
This is addressed in a follow-up PR.

zhangjinpeng1987 commented on src/storage/mod.rs:
```diff
@@ -814,38 +692,11 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
-    ) -> Result<()> {
-        let cmd = commands::ResolveLockLite::new(start_ts, commit_ts, resolve_keys, ctx);
-        self.schedule(cmd, StorageCallback::Boolean(callback))?;
+        // TODO(nrc) only increment one or the other
```
It's a bad idea to increase both.

nrc replied:
This is fixed already in a later commit.

zhangjinpeng1987 commented on src/storage/mod.rs:
```diff
@@ -1428,17 +1278,58 @@ impl<E: Engine, L: LockManager> Storage<E, L> {
+    pub fn from_raw_get(request: &mut RawGetRequest) -> Self {
+        PointGetCommand {
+            ctx: request.take_context(),
+            key: Key::from_raw(request.get_key()),
```
@breeswish So for rawKV we will encode the key? That's weird.

MyonKeminta replied:
we shouldn't

nrc replied:
This code was just moved, not changed. The current code also does this: https://github.com/tikv/tikv/blob/0c9669dd1727b4142d9a5fe248d99a7752e81d17/src/storage/txn/commands.rs#L32.

If this should be changed, then it should be done in a separate PR.

MyonKeminta replied:
I think maybe you misunderstood. `Key::from_raw` is actually encoding a key. For raw kv operations, key should not be encoded and should be written to or read from the engine as-is.

MyonKeminta replied:
Oh I got it. the original code is already like this. It should be incorrect.

MyonKeminta replied:
This problem is introcuded by https://github.com/tikv/tikv/pull/5598 . The unit test uses a wrong way [here](https://github.com/tikv/tikv/pull/5598/files#diff-a2828e4f71806f3e4623d048057803d4R3391) to create the command, so the test didn't fail.
@zhangjinpeng1987 PTAL.

zhangjinpeng1987 commented on components/test_storage/src/sync_storage.rs:
```diff
@@ -185,9 +179,10 @@ impl<E: Engine> SyncTestStorage<E> {
-            .store
-            .commit(ctx, keys, start_ts.into(), commit_ts.into(), cb))
+        wait_op!(|cb| self.store.commit(
+            commands::Commit::new(keys, start_ts.into(), commit_ts.into(), ctx,),
```
Redundant `,`.

nrc commented:
I've rebased and addressed comments. PTAL @zhangjinpeng1987 @MyonKeminta 

Note that I've added another commit which I was going to post separately, since it resolves some of the review comments.

zhangjinpeng1987 commented:
rest LGTM

nrc commented:
> rest LGTM

sorry, I don't see a comment to address

zhangjinpeng1987 commented:
> sorry, I don't see a comment to address

@nrc Please resolve the conflicts, and i will give my LGTM.

nrc commented:
rebased and conflicts fixed. PTAL @zhangjinpeng1987 @MyonKeminta

zhangjinpeng1987 commented:
Build failed.

Little-Wallace commented:
There is an error in 'tests/failpoints/cases/test_storage.rs:42:10'. 
reset LGTM.

nrc commented:
/merge
